src/main/java/org/jenkinsci/plugins/workflow/libs/LibraryConfiguration.java

```diff
@@ -140,14 +204,467 @@ public LibraryCachingConfiguration getCachingConfiguration() {
+                                            scmN.getClass().getName() +
+                                            "': " + scmN.toString());
+                                    }
+                                    if ("hudson.plugins.git.GitSCM".equals(scmN.getClass().getName())) {
+                                        // The best we can do here is accept
+                                        // the first seen SCM (with branch
+                                        // support which we know how to query).
+                                        scm0 = scmN;
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                    }
+
+                    // NOTE: the list of SCMs used by the run does not
+                    // seem trustworthy: if an "inline" pipeline script
+                    // (not from SCM) is used, there is no "relevant"
+                    // branch name to request; however during work on
+                    // JENKINS-69731 I was concerned that the list of
+                    // SCMs might get populated as @Library lines are
+                    // processed and some SCM sources get checked out.
+                    // Experimentally it seems this is not happening,
+                    // and whole pipeline script source is pre-processed
+                    // first (calling this method for many @Library
+                    // lines), and checkouts happen later (populating
+                    // list of SCMs). This is specifically tested by
+                    // checkDefaultVersion_inline_BRANCH_NAME() case.
+                    if (scm0 == null && run instanceof WorkflowRun) {
+                        // This covers both "Multibranch Pipeline"
+                        // and "Pipeline script from SCM" jobs;
+                        // it also covers "inline" pipeline scripts
+                        // but throws a hudson.AbortException since
+                        // there is no SCM attached.
+                        if (logger != null) {
+                            logger.println("defaultedVersion(): inspecting WorkflowRun");
+                        }
+                        try {
+                            WorkflowRun wfRun = (WorkflowRun) run;
+                            scm0 = wfRun.getSCMs().get(0);
+                        } catch (Exception x) {
+                            if (logger != null) {
+                                logger.println("defaultedVersion(): " +
+                                    "Did not get first listed SCM: " +
+                                    x.getMessage());
+                            }
+                        }
+                    }
+
+                    if (scm0 != null) {
+                        // Avoid importing GitSCM and so requiring that
```

Comment:
I suspect adding implementation details of other SCM plugins here is problematic, as support for other SCM plugins means that the code for them has to be maintained here.

What do you think about the following:

Instead add a new abstract method named `hasCheckoutable` (or similar) to https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/scm/SCM.java that then can be implemented/overridden in SCM plugins.

This way no code specific to SCM plugins needs to be added here, it is a bit more work but I think greatly reduces the maintenance burden of this feature inside `pipeline-groovy-lib-plugin`.

Reply:
This would be a neater design indeed. I hoped to avoid large-scale ecosystem changes that would be moot if someone shoots down this PR on ideological basis (security etc). It would also need more expertise and time than I have (and probably testing against SCMs I can't easily make use of other than existing mocks, so I wouldn't see flaws in production).

But sure, fine with me if someone else refactors this better :-D

The goals it should achieve is to have the SCM report which version (branch, tag, ...) the current object refers to if it were to instantiate a workspace, so this feature can learn the `runVersion`. Ideally it would also take care of domain-specific nuances like chopping off the generic git `*/` remote prefix here (so the branch spec `*/main` yields a `main`).

Reply:
Adding a commit to separate the `defaultedVersionSCM()` as a separate large chunk of intimate helper code to make the main `defaultedVersion()` shorter (easier to maintain) and to simplify refactoring later, if it comes to that.